### ISSUE TITLE
`build-container-image` fails with `unableToFind(tool: "swift-autolink-extract")`

### ISSUE THREAD

**Issue by albertodebortoli** [apple/swift-container-plugin#14]
### Description

Thanks for the work done on this plugin @euanh 👏
I'm really excited about it but even though I followed the steps documented, I don't seem to be able to have `build-container-image` working, which is erroring with `unableToFind(tool: "swift-autolink-extract")`.

### Reproduction

I've followed the steps in the [documentation for the plugin](https://github.com/apple/swift-container-plugin/blob/main/Sources/ContainerImageBuilderPluginDocumentation/Documentation.docc/ContainerImageBuilderPlugin.md).

- Downloaded and installed the toolchain `release/6.0` from https://www.swift.org/install/macos/ (snapshot from Sept 11)
- Downloaded and installed the Linux SDK `release/6.0` from  from https://www.swift.org/install/macos/

```bash
swift sdk install swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz --checksum 83a88650cd0675552ce2cf8159c31966fde73418f49493c1644073fffe8be9f4
```

- Confirmed it got installed

```bash
swift sdk list
swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1
```

- Created a demo "hello world" package 

```bash
swift package init --name hello-world --type executable
```

- Added the plugin to the package

```bash
swift package add-dependency https://github.com/apple/swift-container-plugin --from 0.1.0
```

- Added the credentials for my ECR registry to `~/.netrc`

```bash
aws ecr get-login-password --region eu-west-1 // password
```

content of `.netrc`

```
machine <account_id>.dkr.ecr.eu-west-1.amazonaws.com
  login AWS
  password <password>
```

- Ran `build-container-image`

```bash
swift package --swift-sdk swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1 \
  build-container-image --from swift:slim --repository <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
```

resulting in an error `unableToFind(tool: "swift-autolink-extract")`:

```bash
Plugin ‘ContainerImageBuilder’ wants permission to allow all network connections on all ports.
Stated reason: “This command publishes images to container registries over the network”.
Allow this plugin to allow all network connections on all ports? (yes/no) yes
warning: multiple Swift SDKs match ID `swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1` and host triple arm64-apple-macosx15.0, selected one at /Users/alberto/Library/org.swift.swiftpm/swift-sdks/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1/swift-linux-musl
Building for debugging...
[53/53] Applying containertool-tool
Build of product 'containertool' complete! (3.54s)
Building for debugging...
[0/4] Write swift-version--4A847ED0836F2485.txt
[2/5] Emitting module hello_world
<unknown>:0: warning: libc not found for 'aarch64-swift-linux-musl'; C stdlib may be unavailable
error: unableToFind(tool: "swift-autolink-extract")
[3/4] Linking hello-world
error: Build failed
```

I'm not clear on how I get `multiple Swift SDKs match ID` as I only have one.

The same command ran via xcrun results in different logs and error output:

```bash
xcrun --toolchain swift swift package --swift-sdk swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1 build-container-image --from swift:slim --repository 724640160267.dkr.ecr.eu-west-1.amazonaws.com/swift-lambdas
warning: 'swift-container-plugin': /Users/alberto/Desktop/hello-world/.build/checkouts/swift-container-plugin/Package.swift:19:15: warning: 'init(name:defaultLocalization:platforms:pkgConfig:providers:products:dependencies:targets:swiftLanguageVersions:cLanguageStandard:cxxLanguageStandard:)' is deprecated: replaced by ...
<redacted>
31 libsystem_pthread.dylib  0x000000019043f2e4 _pthread_start + 136
32 libsystem_pthread.dylib  0x000000019043a0fc thread_start + 8
[4/48] Emitting module ArgumentParserToolInfo
error: fatalError
➜  hello-world
```

I've tried the same steps using the `6.0-amazonlinux2` image with same results.

I've tried building the HelloWorldVapor example with same results.

Both hello-world and HelloWorldVapor build fine via Xcode.

### Package versions

0.1.0

### Expected behavior

I would expect an image to be generated as shown at the [ServerSide.swift](https://www.serversideswift.info/) conference.

### Environment

macOS version: 15.0
Xcode version installed: 16.0

```bash
swift --version
swift-driver version: 1.115 Apple Swift version 6.0 (swiftlang-6.0.0.9.10 clang-1600.0.26.2)
Target: arm64-apple-macosx15.0

xcrun --toolchain swift swift --version
Apple Swift version 6.0-dev (LLVM fc1036139c5e562, Swift 5bbf725fd958d34)
Target: arm64-apple-macosx15.0
```

**Comment by albertodebortoli** [comment]
On macOS 14.5 and SDK swift-6.0.1-RELEASE_static-linux-0.0.1 I get a different error:

```bash
xcrun --toolchain swift swift package --swift-sdk swift-6.0.1-RELEASE_static-linux-0.0.1 build-container-image --from swift:slim --repository <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
<redacted>
Build of product 'hello-world' complete! (7.04s)
[ContainerImageBuilder] (Error Domain=NSCocoaErrorDomain Code=3840 "Unexpected end of file" UserInfo={NSDebugDescri
error: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional
[ContainerImageBuilder] ption=Unexpected end of file})))
```

**Comment by euanh** [comment]
Hi Alberto.  Thanks for trying the plugin and reporting these problems.   I'm sorry it didn't go smoothly.

I think the problem with `swift-autolink-extract` is probably caused by the OSS toolchain release and SDK you are using on your macOS 15 machine.   This is a problem somewhere in the cross-build.   Could you please try with the most recent releases - 6.0.1 - which don't show this problem on 14.5?   The 6.0.1 toolchain has some important fixes for using the Static SDK.   Please don't forget to update the SDK as well - it's best for the SDK and toolchain to match.

I usually run by setting the `TOOLCHAINS` environment variable, rather than using `xcrun`.  I think the results should be the same, but if you still see this problem after updating, please try:
```
export TOOLCHAINS=$(plutil -extract CFBundleIdentifier raw /Library/Developer/Toolchains/swift-6.0.1-RELEASE.xctoolchain/Info.plist)
swift package --swift-sdk swift-6.0.1-RELEASE_static-linux-0.0.1 build-container-image --from swift:slim --repository <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
```

I've been planning to update the documentation to be more specific about which versions of the tools you need to use.   Unfortunately I haven't quite got it done yet.

**Comment by euanh** [comment]
> On macOS 14.5 and SDK swift-6.0.1-RELEASE_static-linux-0.0.1 I get a different error:
> 
> ```shell
> xcrun --toolchain swift swift package --swift-sdk swift-6.0.1-RELEASE_static-linux-0.0.1 build-container-image --from swift:slim --repository <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
> <redacted>
> Build of product 'hello-world' complete! (7.04s)
> [ContainerImageBuilder] (Error Domain=NSCocoaErrorDomain Code=3840 "Unexpected end of file" UserInfo={NSDebugDescri
> error: dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "The given data was not valid JSON.", underlyingError: Optional
> [ContainerImageBuilder] ption=Unexpected end of file})))
> ```

I think this is a problem talking to ECR.   The hello-world binary looks like it was built correctly, suggesting that the toolchain and SDK you have on this machine are working properly.    You can check by running `file .build/debug/hello-world`, which should show that it's a Linux binary.

The error message isn't very revealing, but the I suspect the problem may be caused by a load balancer in front of the registry which is rejecting the upload.  I'm working on a branch to add support for chunked uploads which might solve this - please check back next week.

In the meantime, if you'd like to verify the example end-to-end, you can run a local registry on your machine like this:
```
docker run -d -p 5050:5000 registry:2
```
Notice that by default the registry listens on port 5000, but that's already in use on recent versions of macOS so we're redirecting it to port 5050.   You can verify that the registry is up and running using `curl` - the `v2/` path (trailing slash is important) should return `{}`:
```
% curl localhost:5050/v2/
{}
```

Then you should be able to build the demo image like this:
```
swift package --swift-sdk x86_64-swift-linux-musl \
    --allow-network-connections all \
    build-container-image --repository localhost:5050/vapor
```

You can then run it like this:
```
docker run -it --rm -p 8080:8080 localhost:5050/vapor:latest
```

**Comment by albertodebortoli** [comment]
Thank you so much for the quick response @euanh!

Believe me if I tell you that I had already tried using a local registry on a port different from 5000 on Docker as you suggested. It was failing a few days ago but didn't try again yesterday when was crafting my initial message here so can't really recall the failure reason. It was likely due to the swift version that stopped me before appreciating that a local registry would work compared to ECR.

I see that with 6.0.1 the binary gets generated on both macOS 14.5 and 15 as the logs in my second message show and as you correctly pointed out.

I can confirm that the fundamental problems for me were 1. not using 6.0.1 and 2. using ECR.

I managed to use the local registry in the end, pull the image from there and push it to ECR in the conventional way.

```bash
docker pull localhost:5050/swift-images:latest
docker tag localhost:5050/swift-images:latest <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
aws ecr get-login-password --region <region> | docker login --username AWS --password-stdin <account_id>.dkr.ecr.eu-west-1.amazonaws.com
docker push <account_id>.dkr.ecr.eu-west-1.amazonaws.com/swift-images
```

I also tried using the [Swift AWS lambda runtime](https://github.com/swift-server/swift-aws-lambda-runtime) with the `swift:6.0.1-amazonlinux2` image as my goal is to have containerized lambdas, but that would be beyond the scope of this issue.

Thank you so much for assisting. I'll check back in the future regarding the chunked upload.
